test/multiverse/suites/logstasher/logstasher_instrumentation_test.rb

```diff
@@ -0,0 +1,123 @@
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+require 'ostruct'
+
+class LogStasherInstrumentationTest < Minitest::Test
+  include MultiverseHelpers
+
+  def setup
+    @written = StringIO.new
+
+    # Give LogStasher's setup method a new place to write logs to, as well as
+    # skip a config check, controller_monkey_patch, that otherwise causes an error
+    LogStasher.setup(OpenStruct.new(logger_path: @written, controller_monkey_patch: false))
+
+    # required for build_logstasher_event method
+    LogStasher.field_renaming = {}
+    @aggregator = NewRelic::Agent.agent.log_event_aggregator
+
+    NewRelic::Agent.instance.stats_engine.reset!
+    NewRelic::Agent.instance.log_event_aggregator.reset!
+  end
+
+  def teardown
+    NewRelic::Agent.instance.stats_engine.reset!
+    NewRelic::Agent.instance.log_event_aggregator.reset!
+  end
+
+  def json_log_hash
+    {
+      :identifier => 'dinosaurs/_dinosaur.html.erb',
+      :name => 'render_partial.action_view',
+      :request_id => '01234-abcde-56789-fghij',
+      'source' => '127.0.0.1',
+      'tags' => [],
+      '@timestamp' => '2024-06-24T23:55:59.497Z'
+    }
+  end
+
+  def test_level_is_recorded
+    in_transaction do
+      LogStasher.build_logstash_event({'level' => :info, 'message' => 'hi there'}, ['log'])
+    end
+    _, events = @aggregator.harvest!
+
+    assert_equal :info, events[0][1]['level']
+    assert_metrics_recorded(%w[Logging/lines/info])
+  end
+
+  def test_logs_without_levels_are_unknown
+    in_transaction do
+      LogStasher.build_logstash_event(json_log_hash, ['log'])
+    end
+    _, events = @aggregator.harvest!
+
+    assert_equal 'UNKNOWN', events[0][1]['level']
+    assert_metrics_recorded(%w[Logging/lines/UNKNOWN])
+  end
+
+  def test_logs_without_messages_are_empty_strings
+    in_transaction do
+      LogStasher.build_logstash_event(json_log_hash, ['log'])
+    end
+    _, events = @aggregator.harvest!
+
+    assert_equal '', events[0][1]['message']
+  end
+
+  def test_attributes_added_to_payload
+    in_transaction do
+      LogStasher.build_logstash_event(json_log_hash, ['log'])
+    end
+    _, events = @aggregator.harvest!
+
+    assert events[0][1]['attributes'].key?(:identifier)
+    assert events[0][1]['attributes'].key?(:name)
+    assert events[0][1]['attributes'].key?('source')
```

Comment:
Nice job checking for compatibility across strings and symbols! Does New Relic turn them all into strings in the end?

Reply:
Too much credit 🫣 This is how the LogStasher reports - some strings, some symbols. Do you think it is worth changing everything to a common class?

Reply:
I think that's a good idea, but I don't think you need to tackle it as part of this PR. Maybe open a separate technical debt ticket?

Reply:
I chose strings because we define string constants for other attributes. But can easily transform into symbols. I know symbols are usually more performant, but I dont think it matters in this case since everytime we iterate over log attributes, they are technically new strings (or symbols). https://github.com/newrelic/newrelic-ruby-agent/pull/2559/commits/3168617c2f006cebb6b94bdb86d9af42ac65c11f

Reply:
This caused some expected Ruby 2.4 errors. I'm reverting for now

Reply:
I think it's fine to leave this as-is. Now that I think about it more, we'll transform everything into strings to send the data as JSON. Sorry for the confusion!